torch/csrc/jit/tensorexpr/ir_simplifier.h

```diff
@@ -0,0 +1,542 @@
+  }
+}
+
+/* Interprets expr as an Immediate and returns the value as type T. */
+template <typename T>
+T immediateAs(const Expr* expr) {
+  T val{0};
+  switch (expr->dtype().scalar_type()) {
+#define TYPE_CASE(Type, Name)                                          \
+  case ScalarType::Name:                                               \
+    if (const Name##Imm* imm = dynamic_cast<const Name##Imm*>(expr)) { \
+      val = imm->value();                                              \
+    } else {                                                           \
+      LOG(FATAL) << "Bad expr: " << *expr << "\n";                     \
+    }                                                                  \
+    break;
+    AT_FORALL_SCALAR_TYPES_AND2(Half, Bool, TYPE_CASE);
+#undef TYPE_CASE
+    default:
+      LOG(FATAL) << "Unsupported datatype: " << expr->dtype();
+  }
+
+  return val;
+}
+
+/* Takes a LinearForm and converts it to Mul + (Add/Sub). */
+const Expr* expandLinearForm(const LinearForm* v, IRMutator* mutator) {
+  const Expr* mul = nullptr;
+  const Expr* A = v->getA();
+  const Expr* B = v->getB();
+  // X can be a LinearForm.
+  const Expr* X = v->getX()->accept_mutator(mutator);
+  // we only really care about 0 and 1, so double should be fine.
+  double Aval = immediateAs<double>(A);
+  double Bval = immediateAs<double>(B);
+
+  // First handle A.
+  if (Aval == 0) {
+    if (Bval == 0) {
+      return getImmediateByType(X->dtype(), 0).node();
+    }
+    return B;
+  } else if (Aval == 1) {
+    mul = X;
+  } else if (Aval == -1) {
+    return new Sub(B, X);
+  } else if (Aval < 0) {
+    // Negate A.
+    ExprHandle zero = getImmediateByType(A->dtype(), 0);
+    Sub* A_Sub = new Sub(zero.node(), A);
+
+    return new Sub(B, new Mul(X, evaluateOp(A_Sub)));
+  } else {
+    mul = new Mul(X, A);
+  }
+
+  if (Bval == 0) {
+    return mul;
+  }
+
+  return new Add(mul, B);
+}
+
+/* Expand any remaining LinearTerms into their component pieces */
+class TermExpander : public IRMutator {
+ public:
+  const Expr* mutate(const LinearForm* v) {
+    return expandLinearForm(v, this);
+  }
+};
+
+/* Simplify the IR by combining arithmetic expressions over a common term.
+ */
+class IRSimplifier : public IRMutator {
+ public:
+  /* Expands lhs and rhs if they are LinearTerms, recurses if they can be folded
+   * into a constant form. */
+  const Expr* expandAndRecurse(
+      const Expr* expr,
+      const Expr* lhs,
+      const Expr* rhs) {
+    if (const LinearForm* lhsLinear = dynamic_cast<const LinearForm*>(lhs)) {
+      lhs = expandLinearForm(lhsLinear, this);
+    }
+    if (const LinearForm* rhsLinear = dynamic_cast<const LinearForm*>(rhs)) {
+      rhs = expandLinearForm(rhsLinear, this);
+    }
+    const Expr* result = newBinaryOpOfType(expr->expr_type(), lhs, rhs, false);
+
+    if (lhs->isConstant() || rhs->isConstant()) {
+      // One wasn't really a linear, we can redo this.
+      return result->accept_mutator(this);
```

Comment:
This recursion call is hard to grasp for me, could you please add some comments on what's going on here? Maybe some examples would help. A brief description of when `LinearForm`s are produced and when they are expanded back to usual expressions might help too.

Reply:
Will do. It's a bit more complicated than it needs to be because we expand what we can in the simplifier pass, then do a second pass to catch the LinearTerms we missed. 

LinearTerms are not constant, but can become constant in the expansion - e.g. if the multiplier term becomes zero. In that case we can continue merging nodes, so we call back into mutate on the newly created op.

Reply:
I was able to eliminate this in the next version (the multivar one).